Fix handling of `file:` module URIs with non-ASCII characters

HT154 opened the pull request:
Resolves #653

bioball reviewed (changes requested):
Thanks for the fix!

We should also fix this in `org.pkl.core.module.FileResolver#listElements(java.net.URI)`, and `org.pkl.core.module.FileResolver#hasElement(java.net.URI)`.

That would handle cases like:

```
read("file:///path/to/日本語.pkl")
```

HT154 commented:
I refactored this into `IoUtils`, adopted it in `FileResolved` and also `StackFrameTransformer`, and added some more test coverage.

I suspect most other usage of `Path.of` in the codebase is at risk of similar issues. Really quite an inconvenient way for this API to work!

HT154 commented on pkl-core/src/test/files/LanguageSnippetTests/output/basic/readGlob.pcf:
```diff
@@ -109,11 +109,15 @@ examples {
     new {
       ["prop:/foo/bar"] = "foobar"
+      ["prop:inputFile"] = "/$snippetsDir/input/basic/readGlob.pkl"
+      ["prop:inputFileDirectory"] = "/$snippetsDir/input/basic/readGlob.pkl"
```
Yup, just forgot to push the updated snippets. Fixed now!

bioball commented on pkl-cli/src/test/kotlin/org/pkl/cli/CliEvaluatorTest.kt:
```diff
@@ -1457,6 +1457,49 @@ result = someLib.x
+      importOne = import("日本語.pkl").readOne
+      importOneFile = import("file:$tempDir/🤬/日本語.pkl").日本語
+      importGlob = import*("./日*.pkl").keys
+      importGlobFile = import*("file:$tempDir/**/*.pkl").keys.map((it) -> it.replaceAll("$tempDir", ""))
```
This will fail on Windows thanks to different filepath separators.

Better to do:

```
val tempDirUri = tempDir.toUri()
```

And use that in the code snippet instead, e.g.

```
read("$tempDirUri/日本語.pkl")
```

HT154 replied:
Ah good catch. I think I've got this sorted, but I have no way to test on windows to be sure.

bioball reviewed (approved):
LGTM pending CI checks.

You need to run `./gradlew spotlessApply` to fix some formatting woes.

HT154 commented:
Hmm I think this may be a legitimate difference in how things are handled on different OSs:
```
Changed content at line 6:
expecting:
  ["at 日本語_error#日本語 (file:///$snippetsDir/input/modules/%E6%97%A5%E6%9C%AC%E8%AA%9E_error.pkl)"]
but was:
  ["at 日本語_error#日本語 (file:///$snippetsDir/input/modules/日本語_error.pkl)"]
```

Is there a way to provide a different expectation or exclude this particular test on Windows? Or would it be best to just move this to `CliEvaluatorTest` where the URI diff can be substituted out of existence?

bioball commented:
Let me look into that one for you; I have access to a Windows machine so it's much easier for me to figure this one out.

translatenix commented on pkl-core/src/main/java/org/pkl/core/util/IoUtils.java:
```diff
@@ -88,6 +88,20 @@ public static URI toUri(String str) throws URISyntaxException {
   }
 
+  /** Converts a URI to a Path, normalizing any non-ASCII characters. */
+  public static Path pathOfURI(URI uri) {
```
```suggestion
  public static Path pathOfUri(URI uri) {
```

bioball commented:
Looks like the issue is that `Path#toUri()` behaves differently on Windows. I added just that one test to the set of excluded tests on Windows.
